### PR TITLE
feat: add PostgreSQL integration

### PR DESCRIPTION
**Pull request by amandahla**
Applicable spec: <link>

### Overview

Maubot requires a database.

Note: This PR is a mashup of code I liked from other charms while working on PostgreSQL integration. I hope you'll find it familiar during your review.

### Rationale

Add PostgreSQL integration as required.

### Juju Events Changes

database-created
endpoints-changed.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

No CH available yet

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -29,40 +39,108 @@ def __init__(self, *args: typing.Any):
...
-            "description": "pebble config layer for httpbin",
+            "description": "pebble config layer for maubot",
             "services": {
                 MAUBOT_SERVICE_NAME: {
                     "override": "replace",
                     "summary": "maubot",
-                    "command": "bash -c \"python3 -c 'import maubot'; sleep 10\"",
+                    "command": "bash -c 'python3 -m maubot -c /data/config.yaml'",
                     "startup": "enabled",
+                    "working-dir": "/data",
```
Comment (javierdelapuente) [inline src/charm.py:143]: Would it be useful (not sure if for other PR better) to use the _daemon_ user instead?

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        relation_data,
+    )
+
+
+def test_maubot_pebble_ready_postgresql_required(harness):
+    """
+    arrange: initialize the testing harness with handle_exec and
+        config.yaml file.
+    act: retrieve the pebble plan for maubot.
+    assert: ensure the maubot pebble plan matches the expectations,
```
Comment (javierdelapuente) [inline tests/unit/test_charm.py:40]: I think this comment is not correct

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        the service is running and the charm is active.
+    """
+    harness.begin()
+
+    harness.container_pebble_ready("maubot")
```
Comment (javierdelapuente) [inline tests/unit/test_charm.py:45]: maybe better to have begin_with_initial_hooks to be more realistic?

**Review by javierdelapuente — COMMENTED** [review]
Nice job. A few comments.

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    Args:
+        harness: harness instance.
+    """
+    relation_data = {
+        "database": "maubot",
+        "endpoints": "dbhost:5432",
+        "password": "somepasswd",  # nosec
+        "username": "someuser",
+    }
+    harness.db_relation_id = (  # pylint: disable=attribute-defined-outside-init
```
[the remark this answers is not in the document]
Comment (amandahla) [reply tests/unit/test_charm.py]: Good catch, fixed.

File: tests/unit/test_charm.py (hunk 2 above)
Comment (amandahla) [reply tests/unit/test_charm.py]: Yes, fixed :)

File: tests/unit/test_charm.py (hunk 3 above)
Comment (amandahla) [reply tests/unit/test_charm.py:44]: Agree, changed.

File: src/charm.py [hunk 5]
```diff
@@ -29,40 +39,108 @@ def __init__(self, *args: typing.Any):
...
         return {
             "summary": "maubot layer",
-            "description": "pebble config layer for httpbin",
+            "description": "pebble config layer for maubot",
             "services": {
                 MAUBOT_SERVICE_NAME: {
                     "override": "replace",
                     "summary": "maubot",
-                    "command": "bash -c \"python3 -c 'import maubot'; sleep 10\"",
+                    "command": "bash -c 'python3 -m maubot -c /data/config.yaml'",
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: No reason at all. Changed.

File: src/charm.py [hunk 6]
```diff
@@ -29,40 +39,108 @@ def __init__(self, *args: typing.Any):
...
+        Returns:
+            postgresql credentials.
+
+        Raises:
+            MissingPostgreSQLRelationDataError: if relation is not found.
+        """
+        relation = self.model.get_relation("postgresql")
+        if not relation or not relation.app:
+            raise MissingPostgreSQLRelationDataError("No postgresql relation data")
+        endpoint = self.postgresql.fetch_relation_field(relation.id, "endpoints")
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py]: good catch, changed it to use only the primary endpoint

File: src/charm.py (hunk 1 above)
Comment (amandahla) [reply src/charm.py:145]: Agree, I'll do this in a further PR.

**Review by javierdelapuente — DISMISSED** [review]

File: src/charm.py [hunk 7]
```diff
@@ -34,60 +44,120 @@ def __init__(self, *args: typing.Any):
...
+            ["cp", "--update=none", "/example-config.yaml", "/data/config.yaml"],
+            ["mkdir", "-p", "/data/plugins", "/data/trash", "/data/dbs"],
+        ]
+        try:
+            for command in commands:
+                process = container.exec(command, combine_stderr=True)
+                process.wait()
+            config_content = str(container.pull("/data/config.yaml", encoding="utf-8").read())
+            config = yaml.safe_load(config_content)
+            config["database"] = self._get_postgresql_credentials()
```
Comment (Thanhphan1147) [inline src/charm.py:78]: I find it strange when there's a try-catch block in this method but we're only handling some errors (`ExecError` + `PathError`) but not all of them (`MissingPostgreSQLRelationDataError` is not caught here but it's caught in the caller method). Maybe we should also catch `MissingPostgreSQLRelationDataError` here and wrap it with another error before raising it to the caller method? what do you think?

**Review by Thanhphan1147 — APPROVED** [review]

**Comment by Thanhphan1147** [comment]
LGTM, some minor comments

**Review by weiiwang01 — APPROVED** [review]

File: src/charm.py (hunk 7 above)
Comment (amandahla) [reply src/charm.py:78]: Its a good point. I'll work on that in the next PR where I need to add NGINX to the Maubot rock.

**Pull request merged**